insonmnia/worker/metrics/metrics.go

```diff
@@ -0,0 +1,194 @@
+		logger: ctxlog.GetLogger(ctx).Named("metrix"),
+	}
+
+	handler.logger.Debug("initializing metrics handler")
+
+	for vendor := range GPUConfig {
+		typeID, err := gpu.GetVendorByName(vendor)
+		if err != nil {
+			return nil, err
+		}
+
+		h, err := gpu.NewMetricsHandler(ctx, typeID)
+		if err != nil {
+			handler.logger.Error("failed to initialize GPU metrics plugin", zap.String("vendor", vendor), zap.Error(err))
+			return nil, err
+		}
+
+		handler.logger.Debug("successfully created GPU handler", zap.String("vendor", vendor))
+		handler.GPUs[typeID] = h
+	}
+
+	return handler, nil
+}
+
+func (m *Handler) Run(ctx context.Context) {
+	go func() {
+		m.logger.Debug("starting metrics collection")
+
+		tk := util.NewImmediateTicker(time.Minute)
+		defer tk.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				m.logger.Warn("context cancelled", zap.Error(ctx.Err()))
+				return
+			case <-tk.C:
+				m.update(ctx)
+			}
+		}
+	}()
+}
+
+func (m *Handler) update(ctx context.Context) {
+	m.logger.Debug("updating hardware metrics")
+	isError := false
+
+	g, err := m.updateGPUMetrics()
+	if err != nil {
+		m.logger.Warn("failed to update GPU metrics", zap.Error(err))
+		isError = true
+	}
+
+	c, err := m.updateCPUMetrics()
+	if err != nil {
+		m.logger.Warn("failed to update CPU metrics", zap.Error(err))
+		isError = true
+	}
+
+	d, err := m.updateDiskMetrics(ctx)
+	if err != nil {
+		m.logger.Warn("failed to update disk metrics", zap.Error(err))
+		isError = true
+	}
+
+	r, err := m.updateRAMMetrics()
+	if err != nil {
+		m.logger.Warn("failed to update RAM metrics", zap.Error(err))
+		isError = true
+	}
+
+	// do not update metrics if some part is failed
+	if !isError {
+		m.mu.Lock()
+		m.lastState = &sonm.WorkerMetricsResponse{
+			GPUMetrics:  g,
+			CpuMetrics:  c,
+			DiskMetrics: d,
+			RamMetrics:  r,
+		}
+		m.mu.Unlock()
+
+		m.logger.Debug("new metrics received") // todo: remove
+	}
+}
+
+func (m *Handler) updateGPUMetrics() ([]*sonm.GPUMetrics, error) {
+	var result []*sonm.GPUMetrics
+
+	for _, h := range m.GPUs {
+		metrics, err := h.GetMetrics()
+		if err != nil {
+			return nil, fmt.Errorf("failed to update GPU metrics: %v", err)
+		}
+
+		result = append(result, metrics...)
+	}
+
+	return result, nil
+}
+
+func (m *Handler) updateCPUMetrics() (*sonm.CPUMetrics, error) {
+	loads, err := cpu.Percent(time.Second, false)
+	if err != nil {
+		return nil, err
+	}
+
+	u := float32(loads[0])
+	return &sonm.CPUMetrics{Utilization: u}, nil
+}
+
+func (m *Handler) updateDiskMetrics(ctx context.Context) (*sonm.DiskMetrics, error) {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	info, err := disk.FreeDiskSpace(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	dm := &sonm.DiskMetrics{
+		Total: info.AvailableBytes,
+		Free:  info.FreeBytes,
+	}
+
+	return dm, nil
+}
+
+func (m *Handler) updateRAMMetrics() (*sonm.RAMMetrics, error) {
+	meme, err := ram.NewRAMDevice()
```

Comment:
meme? something like ![this](https://sd.keepcalm-o-matic.co.uk/i-w600/memes-memes-everywhere.jpg)